README.md
[![Build Status](https://travis-ci.org/sdt/Log-Any-Plugin-ANSIColor.svg?branch=master)](https://travis-ci.org/sdt/Log-Any-Plugin-ANSIColor)
# NAME

Log::Any::Plugin::ANSIColor - Auto-colorize Log::Any logs with Term::ANSIColor

# SYNOPSIS

    use Log::Any::Adapter 'Stderr';     # Choose any adapter that makes sense

    use Log::Any::Plugin;
    Log::Any::Plugin->add('ANSIColor'); # Use the default colorscheme

    # In this or any other module
    use Log::Any qw( $log );

    $log->alert('Call the police!');    # Prints as red on white

# DESCRIPTION

Log::Any::Plugin::ANSIColor automatically applies ANSI colors to log messages depending on the log level.

# USAGE

Adding the plugin with no extra arguments gives the default colorscheme. Info and notice messages have no special coloring.

    Log::Any::Plugin->add('ANSIColor');

Specify some colors to completely replace the default colorscheme. Only the specified colors are applied.

    Log::Any::Plugin->add('ANSIColor',
            error   => 'white on_red',
            warning => 'black on_yellow',
    );

Use `default => 1` to include the default colorscheme with customisations. Default colors can be switched off by specifying `'none'` as the color.

    Log::Any::Plugin->add('ANSIColor',
            default => 1,               # use default colors
            error   => 'white on_red',  # override error color
            warning => 'none',          # turn off warning color
    );

Valid colors are any strings acceptable to `colored` in [Term::ANSIColor](https://metacpan.org/pod/Term::ANSIColor).
eg. `'blue'` `'bright_red on_white`

# LICENSE

Copyright (C) Stephen Thirlwall.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Stephen Thirlwall <sdt@cpan.org>


Changes
Revision history for Perl extension Log-Any-Plugin-ANSIColor

{{$NEXT}}

0.02 2019-11-07T05:50:17Z

    - don't colorize references arguments - thanks Yanick Champoux, and sorry for the slow turnaround
    - colorize alias methods the same as their primary counterparts

0.01 2017-09-22T01:01:42Z

    - original version



lib/Log/Any/Plugin/Color.pm
package Log::Any::Plugin::Color;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.01";

use Log::Any::Plugin::Util  qw( get_old_method set_new_method );
use Term::ANSIColor         qw( colored colorvalid );

our %default = (
    emergency  => 'red on_white',
    alert      => 'red on_white',
    critical   => 'red on_white',
    error      => 'red',
    warning    => 'yellow',
    debug      => 'cyan',
    trace      => 'blue',
);

sub install {
    my ($class, $adapter_class, %color_map) = @_;

    if ((delete $color_map{default}) || (keys %color_map == 0)) {
        # Copy the default colors, leaving any the user has specified.
        for my $method (keys %default) {
            $color_map{$method} ||= $default{$method};
        }
    }

    for my $method_name (Log::Any->logging_methods) {
        if (my $color = delete $color_map{$method_name}) {
            # Specifying none as the color name disables colorisation for that
            # method.
            next if $color eq 'none';

            if (!colorvalid($color)) {
                warn "Invalid color name \"$color\" for $method_name";
                next;
            }

            my $old_method = get_old_method($adapter_class, $method_name);
            set_new_method($adapter_class, $method_name, sub {
                my $self = shift;
                $self->$old_method(colored([$color], @_));
            });
        }
    }

    if (my @remainder = sort keys %color_map) {
        warn 'Unknown logging methods: ', join(', ', @remainder);
    }
}

1;
__END__

=encoding utf-8

=head1 NAME

Log::Any::Plugin::Color - Auto-colorize logs using Term::ANSIColor

=head1 SYNOPSIS

    use Log::Any::Plugin;

    Log::Any::Plugin->add('Color', default => 1, debug => 'blue on_white');

=head1 DESCRIPTION

Log::Any::Plugin::Color applies ANSI colors to log messages depending on the log level.

=head1 LICENSE

Copyright (C) Stephen Thirlwall.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Stephen Thirlwall E<lt>sdt@cpan.orgE<gt>

=cut



lib/Log/Any/Plugin/ANSIColor.pm
package Log::Any::Plugin::ANSIColor;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.01";

use Log::Any::Plugin::Util  qw( get_old_method set_new_method );
use Log::Any::Adapter::Util qw(
    log_level_aliases logging_aliases logging_methods
);
use Term::ANSIColor         qw( colored colorvalid );

our %default = (
    emergency  => 'bold magenta',
    alert      => 'magenta',
    critical   => 'bold red',
    error      => 'red',
    warning    => 'yellow',
    debug      => 'cyan',
    trace      => 'blue',
);

sub install {
    my ($class, $adapter_class, %color_map) = @_;

    if ((delete $color_map{default}) || (keys %color_map == 0)) {
        # Copy the default colors, leaving any the user has specified.
        for my $method (keys %default) {
            $color_map{$method} ||= $default{$method};
        }
    }

    my %aliases = log_level_aliases();
    for my $alias (keys %aliases) {
        my $method = $aliases{$alias};

        $color_map{$alias} = $color_map{$method}
            if exists $color_map{$method};
    }

    for my $method_name ( logging_methods(), logging_aliases() ) {
        if (my $color = delete $color_map{$method_name}) {
            # Specifying none as the color name disables colorisation for that
            # method.
            next if $color eq 'none';

            if (!colorvalid($color)) {
                warn "Invalid color name \"$color\" for $method_name";
                next;
            }

            my $old_method = get_old_method($adapter_class, $method_name);
            set_new_method($adapter_class, $method_name, sub {
                my $self = shift;
                $self->$old_method(
                    # Colorise non-ref arguments, leave ref args alone.
                    map { ref $_ ? $_ : colored([$color], $_) } @_
                );
            });
        }
    }

    if (my @remainder = sort keys %color_map) {
        warn 'Unknown logging methods: ', join(', ', @remainder);
    }
}

1;
__END__

=encoding utf-8

=head1 NAME

Log::Any::Plugin::ANSIColor - Auto-colorize Log::Any logs with Term::ANSIColor

=head1 SYNOPSIS

    use Log::Any::Adapter 'Stderr';     # Choose any adapter that makes sense

    use Log::Any::Plugin;
    Log::Any::Plugin->add('ANSIColor'); # Use the default colorscheme

    # In this or any other module
    use Log::Any qw( $log );

    $log->alert('Call the police!');    # Prints as red on white

=head1 DESCRIPTION

Log::Any::Plugin::ANSIColor automatically applies ANSI colors to log messages depending on the log level.

For example, with the default colorscheme, C<error> logs are red, C<warning> logs are yellow.

If a given log level has no coloring, the original log method is left intact, and incurs no overhead.

=head1 USAGE

Adding the plugin with no extra arguments gives the default colorscheme.

    Log::Any::Plugin->add('ANSIColor');

Note that C<info> and C<notice> messages have no special coloring in the default colorscheme.


Specify some colors to completely replace the default colorscheme. Only the specified colors are applied.

    Log::Any::Plugin->add('ANSIColor',
            error   => 'white on_red',
            warning => 'black on_yellow',
    );

Use C<< default => 1 >> to include the default colorscheme with customisations. Default colors can be switched off by specifying C<'none'> as the color.

    Log::Any::Plugin->add('ANSIColor',
            default => 1,               # use default colors
            error   => 'white on_red',  # override error color
            warning => 'none',          # turn off warning color
    );

Valid colors are any strings acceptable to C<colored> in L<Term::ANSIColor>.
eg. C<'blue'> C<'bright_red on_white>

=head1 LICENSE

Copyright (C) Stephen Thirlwall.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Stephen Thirlwall E<lt>sdt@cpan.orgE<gt>

=cut

